### PR TITLE
feat: Prune Paths + Fast ABI

### PR DESCRIPTION
**Pull request by dekz**
## Description

TKR-24

[Simbot sells](https://metabase.spaceship.0x.org/dashboard/186?run_id=jacob-fast-abi--mainnet&adjusted_amount_bought_win_tolerance__bps_=0.5)
### Prune paths

Prune paths where the rate (`bestRate`) when selling a fractional amount is less than the adjusted complete rate of selling the entire amount (`adjustedCompleteRate`) to the best source (sorted by `adjustedCompleteRate`).

Since selling anything to this path wouldn't result in any overall improvement. We can cut down on the paths to compare.

We should see no regressions, only improved response times (there is the small possibility of improved prices as we have more time to search over less paths)

### Fast ABI

Use an ABI encoder/decoder which is more performant than our AbiEncoder (though lacks the transaction gas optimization).
 
## Types of changes

```
4000 USDC buy 
  before: 26,
  after: 21
400,000 USDC buy (drops Uni+Sushi, goes hard on Curves + MakerPSM)
  before: 21,
  after: 9
1 ETH sell (drops CryptoCom)
  before: 30,
  after: 26
100 ETH sell (drops Linkswap, Mooni, CryptoCom, Balancer)
  before: 25,
  after: 6
```

```
  MarketOperationUtils tests
    optimization
      100 routes
old: 1759ms 56.85 routes/s
new: 199ms 502.51 routes/s
        ✓ DEX QUOTES BNB/BUSD SMALL (1963ms)
old: 6907ms 14.48 routes/s
new: 370ms 270.27 routes/s
        ✓ DEX QUOTES BNB/BUSD LARGE (7283ms)
old: 1661ms 60.2 routes/s
new: 372ms 268.82 routes/s
        ✓ DEX QUOTES DAI/USDC LARGE (2041ms)
old: 22057ms 4.53 routes/s
new: 18470ms 5.41 routes/s
        ✓ DEX QUOTES ETH/DAI LARGE (40531ms)
old: 22361ms 4.47 routes/s
new: 19078ms 5.24 routes/s
NEW IS LARGER BY 0.005474% 45571513355722468522 units
        ✓ DEX QUOTES ETH/DAI LARGE (FREE GAS) (41442ms)
old: 2285ms 43.76 routes/s
new: 2218ms 45.09 routes/s
        ✓ DEX QUOTES ETH/DAI SMALL (4509ms)
old: 2141ms 46.71 routes/s
new: 2067ms 48.38 routes/s
        ✓ DEX QUOTES ETH/USDC SMALL (4213ms)


  7 passing (2m)
```

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 1]
```diff
@@ -48,6 +70,17 @@ export function fillsToSortedPaths(
...
 
+export function fillsToSortedPaths2(
+    fills: Fill[][],
+    side: MarketOperation,
+    targetInput: BigNumber,
+    opts: PathPenaltyOpts,
+): Path[] {
+    const paths = fills.map(singleSourceFills => Path.create(side, singleSourceFills, targetInput, opts));
+    const sortedPaths = paths.sort((a, b) => b.adjustedCompleteRate().comparedTo(a.adjustedCompleteRate()));
+    return sortedPaths.filter(p => p.minRate().isGreaterThanOrEqualTo(sortedPaths[0].adjustedCompleteRate()));
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:81]: `minRate()` is unadjusted tho

**Review by dorothy-zbornak — APPROVED** [review]

File: packages/asset-swapper/src/utils/market_operation_utils/path.ts [hunk 2]
```diff
@@ -39,6 +39,7 @@ export class Path {
     public sourceFlags: number = 0;
     protected _size: PathSize = { input: ZERO_AMOUNT, output: ZERO_AMOUNT };
     protected _adjustedSize: PathSize = { input: ZERO_AMOUNT, output: ZERO_AMOUNT };
+    private _firstFill?: Fill;
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/path.ts:42]: can we just use `this.fills[0]`?

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 3]
```diff
@@ -48,6 +49,16 @@ export function fillsToSortedPaths(
     return sortedPaths;
 }
 
+// Remove paths which have no impact on the optimal path
+export function reducePaths(sortedPaths: Path[]): Path[] {
+    if (sortedPaths.length === 0) {
+        return sortedPaths;
+    }
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:56]: I think this is unnecessary, `[].filter(() => false)` returns `[]` anyway

**Review by mzhu25 — APPROVED** [review]
looks good to me, just a couple of nits

File: packages/asset-swapper/src/utils/market_operation_utils/path.ts (hunk 2 above)
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/path.ts]: adjusted to use the BEST fill (we have some ETH2Dai tests which apparently the rate can get better with more fills somehow)

**Review by dorothy-zbornak — APPROVED** [review]

File: packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts [hunk 4]
```diff
@@ -44,10 +45,33 @@ export function fillsToSortedPaths(
     opts: PathPenaltyOpts,
 ): Path[] {
     const paths = fills.map(singleSourceFills => Path.create(side, singleSourceFills, targetInput, opts));
-    const sortedPaths = paths.sort((a, b) => b.adjustedCompleteRate().comparedTo(a.adjustedCompleteRate()));
+    const sortedPaths = paths
+        .sort((a, b) => b.adjustedCompleteRate().comparedTo(a.adjustedCompleteRate()))
+        .filter(a => a.adjustedCompleteRate().isGreaterThanOrEqualTo(0));
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/asset-swapper/src/utils/market_operation_utils/path_optimizer.ts:50]: image

**Pull request merged**